// html/index.js
module.exports = html

var persist = require('persist');

function html(tag, attributes, css, content) {
    if (!/^[a-zA-Z0-9]+$/.test(tag)) {
        return '<span style="color:red">Invalid tag name [TODO: insert tag name]</span>'
    }
    
    var processed_content = content;
    
    //  Apply necesary code and remove script element
    function onload() {
        var id = ID;
        var css = CSS;
        var attr = ATTR;
        var element = $("#" + id);
        var script = $("#" + id + 'script');
        function apply_css(name, value) {
            if (value.type == 'string' || value.type == 'number') {
                element.css(name, value.data);
            }
        }
        function apply_attribute(name, value) {
            if (value.type == 'string' || value.type == 'number') {
                element.attr(name, value.data);
            }
        }
        $(function () {
            for (var field in css) {
                apply_css(field, css[field]);
            }
            for (var field in attr) {
                apply_attribute(field, attr[field])
            }
            console.log(':)', element, css, attr);

        });
        //element.removeAttr('id');
        //script.remove();
    }

    function serialize_fields(from, into) {
        for (var field in from) {
            try {into[field] = persist.serializable_from_instance(from[field]);}
            catch (error) {} // should never error since persistant objects can only have serializable values
        }
    }
    var id = (Math.random() + '').slice(2);
    var new_css = {};
    serialize_fields(css, new_css)
    css = new_css;
    
    var new_attr = {};
    serialize_fields(attributes, new_attr);
    attributes = new_attr;
 
    onload = '(' + onload.toString()
        .replace('CSS', JSON.stringify(css))
        .replace("ATTR", JSON.stringify(attributes))
        .replace("ID", '"'  + id + '"') + ')()';
    var script = '<script id="' + id + 'script" type="text/javascript">' + onload + '</script>'
    return '<' + tag + ' id="' + id + '">' + processed_content + '</' + tag + '>' + script
};
