// html/index.js
module.exports = html;

var persist = require('persist');
var references = require('references');
var util = require('util');

function html(tag, attributes, content) {
    if (!/^[a-zA-Z0-9]+$/.test(tag)) {
        return '<span style="color:red">Invalid tag name [TODO: insert tag name]</span>';
    }

    var scripts = '';
    
    var processed_content;
    function process_content(content) {
        try {
            if (persist.is(content)) {
                var render = content.render;
                if (persist.get_type(render) == 'function') {
                    processed_content = persist.serializable_from_instance(render());
                }
                else if (persist.get_type(render) == 'reference') {
                    processed_content = persist.serializable_from_instance(render);
                }
                else {
                    processed_content = {type : 'string', data : 'invalid renderer for an object'};
                }
                return insert_content(processed_content);
            }
            else if (util.isArray(content)) {
                var scripts = "";
                for (var i=0; i<content.length; i++) {
                    scripts += process_content(content[i]);//insert_content(persist.serializable_from_instance(content[i]));
                }
                return scripts;
            }
            else {
                processed_content = persist.serializable_from_instance(content);
                return insert_content(processed_content);
            }
        }
        catch (error) {
            //  TODO: error content...
            return insert_content(persist.serializable_from_instance(error));
        }
    }

    var id = (Math.random() + '').slice(2);

    function insert_content(processed_content) {
        //  Apply necesary code and remove script element
        function client() {
            var id = ID;
            var script_id = SCRIPT_ID;
            //  valid HTML within strings, so need to encode and decode front-end
            var content = JSON.parse(decodeURIComponent(CONTENT));
            var element = $("#" + id);
            var script = $("#" + script_id);
            $(function () {
                bind_html(element, content);
            });
            script.remove();
        }

        var script_id = (Math.random() + "").substring(2);
        var client_source = '(' + client.toString()
            .replace("CONTENT", '"' + encodeURIComponent(JSON.stringify(processed_content)) + '"')
            .replace("ID", '"'  + id + '"')
            .replace("SCRIPT_ID", '"' + script_id + '"') + ')()';
        return '<script id="' + script_id + '" type="text/javascript">' + client_source + '</script>';
    };

    function apply_attributes(attributes) {
    //  Apply necesary code and remove script element
        function client() {
            //  valid HTML within strings, so need to encode and decode front-end
            var attr = JSON.parse(decodeURIComponent(ATTR));
            $(function () {
                var field;
                for (field in attr) {
                    bind_attribute(element, field, attr[field]);
                }
            });
            script.remove();
        }

        function serialize_attributes(from, into) {
            try {
                for (var field in from) {
                    //  style value may be an object for css reasons
                    if (field == 'style') {
                        //  TODO: handle if from.style is a reference
                        into.style = {};
                        serialize_attributes(from.style, into.style);
                    }
                    else {
                        // TODO: somehow warn object not valid
                        try {into[field] = persist.serializable_from_instance(from[field]);}
                        catch (error) {}            }
                }
            }
            catch (error) {
                //  TODO: somehow warn object not valid
            }
        }

        var new_attr = {};
        serialize_attributes(attributes, new_attr);
        attributes = new_attr;
        var script_id = (Math.random() + "").substring(2);
        var client_source = '(' + client.toString().replace("ATTR", '"' + encodeURIComponent(JSON.stringify(attributes)) + '"') + ')()';
        return '<script id="' + script_id + '" type="text/javascript">var element = $("#' + id + '");\nvar script = $("#' + script_id + '");\nelement.removeAttr("id");\n' + client_source + '</script>';
    }

    var script_id = (Math.random() + "").substring(2);
    var r = '<' + tag + ' id="' + id + '"></' + tag + '>' + process_content(content) + apply_attributes(attributes);
    return r;
}
