// files/index.js
//  get a signed url for a file upload
var url = require('url');
var aws = require('aws');
var s3 = aws.s3;

var Bucket = 'persist';

module.exports.server = function (req, res, next) {
    if (req.path.length > 1) {
        //  get the requested file
        var Key = req.path.slice(1);
        var params = {
            Bucket : Bucket,
            Key : Key,
            Expires : 60
        };
        s3.getSignedUrl('getObject', params, function (err, URL) {
            if (err) res.status(500).end(err + '');
            else     res.redirect(URL);
        });
    }
    else if (req.path.length == 1){
        var ContentType = req.query.ContentType;
        var Key = Date.now() + (Math.random() + '').substring(1);
        var params = {
            Bucket : Bucket,
            Key : Key,
            ACL : 'public-read',
            ContentType : ContentType,
            Expires : 60
        };
        s3.getSignedUrl('putObject', params, function (err, URL) {
            if (err) {
                res.statusCode = 500;
                res.end(err + '');
            }
            else {
                res.statusCode = 200;
                res.end(JSON.stringify({
                    upload : URL,
                    download  : '/' + Key
                }));
            }
        });
    }
    else next();
};

var file_url = new Map();

module.exports.file = function File(url) {
    function stream() {
        //  return file streaming object
    }

    function get() {
        //  return promise and collect entire file through stream
    }
    
    this.stream = stream;
    this.get    = get;
};

module.exports.url = function(file) {
    if (file_url.has(file)) return file_url.get(file);
    else throw new Error('cannot get url of non-file object');
};
