// files/server.js
//  get a signed url for a file upload
var url = require('url');
var aws = require('aws');
var s3 = aws.s3;

var Bucket = 'persist';

module.exports = function (req, res, next) {
    if (req.path.length > 1) {
        //  get the requested file
        var Key = req.path.slice(1);
        var params = {
            Bucket : Bucket,
            Key : Key,
            Expires : 60
        };
        operation = 'getObject';
        if (req.method == 'HEAD') {
            operation = 'headObject'
        }
        s3.getSignedUrl(operation, params, function (err, URL) {
            if (err) res.status(500).end(err + '');
            else     res.redirect(URL);
        });
    }
    else if (req.path.length == 1) {
        var ContentType = req.query.ContentType;
        var Key = Date.now() + (Math.random() + '').substring(1);
        var params = {
            Bucket : Bucket,
            Key : Key,
            ACL : 'private',
            Expires : 60
        };
        if (ContentType.length) params.ContentType = ContentType;
        else params.ContentType = 'binary/octet-stream';
        s3.getSignedUrl('putObject', params, function (err, URL) {
            if (err) {
                res.statusCode = 500;
                res.end(err + '');
            }
            else {
                res.json({
                    upload : URL,
                    filename : Key
                });
            }
        });
    }
    else next();
};
