// persist/create.js
module.exports = create;

var get_time = require('./get_time');
var containers = require('./containers');
var lock = require('./lock');
var meta = require('./meta');
var records = require('./records');
var serializable_from_instance = require('./serializable_from_instance');

function create(type, id, nobackup) {

    //  creates persistant object of type with a new id
    if (id === undefined) id = get_time() + '';
    else if (typeof id != 'string') {
        //console.log(id);
        throw new Error('id must be a string');
    }

    var proxy = containers.create(type, id);
    lock(id, function (err, res) {
        if (err) console.log(err);
    });
    meta.get(proxy).loaded = true;
    meta.get(proxy).backup = true;

    //  write in the type information
    if (type) {
        records.save({
            time : get_time(),
            id : id,
            operation : 'set',
            meta : true,
            field : 'type',
            value : serializable_from_instance(type)

        });
        meta.get(proxy).type = type;
    }
    else {
        throw new Error('Must create persistant object with a type');
    }

    //  don't need to delete, because the first set
    //  to type is never saved on the instance itself
    //      only on the persistant representation
    //delete proxy.type;

    if (nobackup) {
        meta.get(proxy).backup = false;
    }

    return proxy;
}
