// persist/create.js
var get_time = require('./get_time');
var get_persistant_proxy = require('./get_persistant_proxy');
var lock = require('./lock');
var meta = require('./meta');
var save_record = require('./save_record');

module.exports = create;

var valid_types = {
    'hashmap' : true,
    'list'    : true
};

function create(type, id) {
    if (valid_types[type] === undefined) {
        throw new Error('"' + type + '" is not a valid type');
    }

    //  creates persistant object of type with a new id
    if (id === undefined) id = get_time() + '';
    else if (typeof id != 'string') {
        throw new Error('id must be a string');
    }

    var proxy = get_persistant_proxy(id);
    save_record({
        time      : get_time(),
        id        : id,
        operation : 'create',
        field     : 'type',
        type      : type,
        value     : type
    });
    lock(id, function (err, res) {
        if (err) console.log(err);
    });
    meta.get(proxy).loaded = true;
    meta.get(proxy).type = type;
    return proxy;
}
