// persist/reference.js
var persist = require('persist');
var net = require('net');

var meta = new Map();

//  id to to persistant object
//  internal refernce is to data inside of object with id
function create(id, internal_reference) {
    function keys(target) {return [];}
    function set (target, field, val, receiver) {}
    function get (target, name, receiver) {
        return create(id, name);
    }

    var proxy = Proxy.create({
        getOwnPropertyDescriptor : function (target, name) {},
        getOwnPropertyNames      : function (target) {},
        delete                   : function (name) {},
        freeze                   : function () {},
        seal                     : function () {},
        preventExtensions        : function () {},
        has                      : function () {},
        hasOwn                   : function () {},
        enumerate                : function (target) {},
        get                      : get,
        set                      : set,
        keys                     : keys,
        apply                    : function () {},
        construct                : function () {}
    });
    var promise = new Promise(function (resolve, reject) {
        var reference_process = persist.host(id);
        function retrieve(processID) {
            //  use TCP socket to connect to server with object
            var client = net.connect({
                port : 8888,
                host : processID
            }, function () {
                client.write(JSON.stringify({
                    id : id,
                    internal : internal_reference
                }));
            });
            client.on('data', function (data) {
                //  TODO: convert data into real instance of transmitted object
                var info = JSON.parse(data);
                resolve(info);
            });
        }
        reference_process.then(retrieve, reject);
    });
    meta.set(proxy, {
        promise : promise,
        id    : id
    });
    return proxy;
}

function path(ref) {
    if (ref_path.has(reference)) return meta.get(ref).path;
    else throw new Error('cannot get path for non reference');
}

function resolve(reference) {
    return meta.get(reference).promise;
}

function is(reference) {
    return meta.has(reference);
}

module.exports.create  = create;
module.exports.path    = path;
module.exports.resolve = resolve;
module.exports.is      = is;
