// persist/send_to_watching.js
var hash_reference = require('./hash_reference');
var watched_references = require('./watched_references');
var serializable_from_instance = require('./serializable_from_instance');
var get_time = require('./get_time');

var updates_pending = {};

function send_to_watching(context, reference, value) {
    var hashed_ref = hash_reference(reference);
    var watching = watched_references[hashed_ref];
    if (watching) {
        //  overwrite existing update function if exists
        updates_pending[hashed_ref] = true;
        process.nextTick(function () {
            //  only execute update function once per process tick
            if (updates_pending[hashed_ref]) {
                delete updates_pending[hashed_ref];
                var serializable = serializable_from_instance(value, reference);
                //  TODO: send original time along
                var update = JSON.stringify({
                    type : 'update',
                    reference : reference,
                    time : get_time(),
                    value : serializable,
                    token : get_time()
                });
                for (var i=0; i<watching.length; i++) {
                    //  if context is also watching this reference, it does not need to receive the update
                    if (context != watching[i] && watching[i].open) {
                        watching[i].respond.call(watching[i], '"""'  + update);
                    }
                }
            }
        });
    }
}

module.exports = send_to_watching;
