// modules/index.js
var path = require('path');
var fs = require('fs');

var loaded = {};
var meta = new Map();

//  check for new modules every minute
var check_time = 60000;

var available = {};

function refresh_available() {
    fs.readdir(__dirname + '/node_modules', function (err, res) {
        if (err) {}
        else {
            available = {};
            for (var i=0; i<res.length; i++) {
                available[res[i]] = true;
            }
        }
    })
    setTimeout(refresh_available, check_time);
}

refresh_available();

module.exports.is = function (value) {
    return meta.get(value);
}

module.exports.identify = function (value) {
    return meta.get(value).name;
}

module.exports.serializable = function (value) {
    return {
        name : meta.get(value).name
    };
}

module.exports.loader = Proxy.create({
    getOwnPropertyDescriptor : function (target, name) {
        return loaded[name];
    },
    getOwnPropertyNames : function (target) {
        return Object.keys(loaded);
    },
    delete : function (name) {
        //  TODO: any unload behavior
        proxy[name] = undefined;
    },
    freeze : function () {},
    seal : function () {},
    preventExtensions : function () {},
    has : function () {},
    hasOwn : function () {},
    get : function (target, name, receiver) {
        //  TODO: restrict where can load from
        if (loaded[name] === undefined && available[name]) {
                loaded[name] = require(name);
                meta.set(loaded[name], {name : name});
        }
        
        if (available[name]) {
            return loaded[name];
        }
        else {
            delete loaded[name];
        }
    },
    set : function () {
        throw new Error('You may not set properties on the module object');
   },
    enumerate : function (target) {
        return Object.keys(loaded);
    },
    keys : function (target) {
        return Object.keys(loaded);
    },
    apply : function () {},
    construct : function () {}
});
