// modules/index.js
var path = require('path');

var loaded = {};

module.exports = Proxy.create({
    getOwnPropertyDescriptor : function (target, name) {
        return loaded[name];
    },
    getOwnPropertyNames : function (target) {
        return Object.keys(loaded);
    },
    delete : function (name) {
        //  TODO: any unload behavior
        proxy[name] = undefined;
    },
    freeze : function () {},
    seal : function () {},
    preventExtensions : function () {},
    has : function () {},
    hasOwn : function () {},
    get : function (target, name, receiver) {
        //  TODO: restrict where can load from
        if (loaded[name] === undefined) {
            try {
                var resolved = path.resolve(__dirname + '/' + name);
                var root = resolved.slice(0, __dirname.length);
                //  only load module if looking into this dirname
                if (root == __dirname) {
                    loaded[name] = require(resolved);
                }
            }
            catch (error) {
                //  simply leave undefined
            }
        }
        return loaded[name];
    },
    set : function () {
        throw new Error('You may not set properties on the module object');
   },
    enumerate : function (target) {
        return Object.keys(loaded);
    },
    keys : function (target) {
        return Object.keys(loaded);
    },
    apply : function () {},
    construct : function () {}
});
