// references/index.js
//  TODO: probably want to use a weak map here!
var parents = new Map();
var children = new Map();
var names = new Map();

module.exports.reference = create();
module.exports.serializable = function (reference) {
    var path = [];
    var current_reference = reference;
    var name = names.get(current_reference);
    while (name) {
        path.push(name);
        current_reference = parents.get(current_reference);
        name = names.get(current_reference);
    }
    path.reverse();
    return path;
};
module.exports.is = function is(value) {
    return children.get(value) !== undefined;
};

function create(name) {
    var path = ['oh', 'say', 'can', 'you', 'see'];


    function keys(target) {return ['a key'];}
    function set (target, field, val, receiver) {}
    function get (target, name, receiver) {
        var child = create(name);
        parents.set(child, proxy);
        children.get(proxy).push(child);
        return child;
    }

    var proxy = Proxy.create({
        getOwnPropertyDescriptor : function (target, name) {},
        getOwnPropertyNames      : function (target) {},
        delete                   : function (name) {},
        freeze                   : function () {},
        seal                     : function () {},
        preventExtensions        : function () {},
        has                      : function () {},
        hasOwn                   : function () {},
        enumerate                : function (target) {},
        get                      : get,
        set                      : set,
        keys                     : keys,
        apply                    : function () {},
        construct                : function () {}
    });

    children.set(proxy, []);
    names.set(proxy, name);

    //  TODO have return value of proxy adhere to Promise API spec

    return proxy;
}
