// environment/index.js
var references = require('references');
var files      = require('files');
var promises   = require('promises');
var html       = require('html');
var persist    = require('persist');
var clients    = require('clients');
var collections = require('persist/collections')
var modules = require('modules');

var exposed_modules = {
    'timers' : true //  going to be the set timeout and set inteval interface
};

var loaded = {};

module.exports = {
    Hashmap : collections.Hashmap,
    List : collections.List,
    Set : collections.Set,
    Promise : collections.Promise,
    console : console,
    client : clients.client,
    module : modules,
    unload : function unload(name) {
        var m = loaded[name];
        if (m) m.unload();
        else throw new Error('"' + name + '" is not loaded');
    },
    reference : references.reference,
    resolve : references.resolve,
    File : files.File,
    html : html,
    type : function () {
        return persist.get_type.apply(this, arguments);
    }
//    setTimeout : setTimeout,  need way to reference and stop outstanding timeouts and intervals
//    setInterval : setInterval
}
