// environment/index.js
var references = require('references');
var files      = require('files');
var promises   = require('promises');
var html       = require('html');
var persist    = require('persist');
var clients    = require('clients');
var modules = require('modules');

//var exposed_modules = {
//    'timers' : true //  going to be the set timeout and set inteval interface
//};

module.exports = {
//    profile : require('profile'), //  everything is profiled anyway...
    Item    : function () {return persist.create('hashmap');},
    List    : function () {return persist.create('list');},
    Promise : function () {return persist.create('promise');},
    File : files.File,
    Client : clients.Client,
//    Set : collections.Set,
//    console : console,
//    client : clients.client,
    module : modules.loader,
    external : references.reference,
    resolve : references.resolve,
    type : persist.get_type
//    setTimeout : setTimeout,  need way to reference and stop outstanding timeouts and intervals
//    setInterval : setInterval
//    html : html,
}
