// clients/index.js
module.exports.client = client;
module.exports.Client = Client;

module.exports.is = function (test) {
    return meta.get(test) !== undefined;
}
module.exports.serializable = function (client) {
    var serialized = {};
    serialized.language = meta.get(client).language;
    serialized.script = meta.get(client).script;
    return serialized;
}

module.exports.unserialize = function (data) {
    return Client(data.language, data.script);
}

var persist = require('persist');
var references = require('references');

var meta = new Map();

function Client(language, script) {
    if (typeof language !== 'string') {
        throw new Error('must specify language as a string');
    }
    if (typeof script !== 'string') {
        script = '';
    }
    var hash = {
        language : language,
        script : script
    };
    var proxy = Proxy.create({
        getOwnPropertyDescriptor : function (target, name) {},
        getOwnPropertyNames : function (target) {
            return [];
        },
        delete : function (name) {},
        freeze : function () {},
        seal : function () {},
        preventExtensions : function () {},
        has : function () {},
        hasOwn : function () {},
        get : function (target, name, receiver) {
            return hash[name];
        },
        set : function (target, field, val, receiver) {
            if (field == 'script' && typeof val === 'string') {
                hash[name] = val;
            }
        },
        enumerate : function (target) {
            return [];
        },
        keys : function (target) {
            return [];
        },
        apply : function () {},
        construct : function () {}
    });
    meta.set(proxy, hash);
    return proxy;
}

var clientized = function () {
    ENVIRONMENT 
    var fn = CLIENT;
    fn();
};


//  returns a function that can be executed on the client and
//  has access to remote functions
function client(client_fn, remote_fns) {
    //  TODO: ensure all remote functions are references to functions

    var environment = '';

    for (var fn in remote_fns) {
        //  TODO: return promise and execute remote function
        //  TODO: add code to add the arguments to the function to the
        environment += 'var reference = ' + JSON.stringify(references.serializable(remote_fns[fn])) + '\n'
                    +  '    reference.push({type : "call", name : "*"});\n'
                    +  '    var ' + fn + ' = function () {\n'
                    +  '        var args = [];\n'
                    +  '        reference[reference.length-1].arguments = args;\n'
                    +  '        for (var i=0; i<arguments.length; i++) args.push(serializable_from_instance(arguments[i]));\n'
                    +  '        resolve_reference(reference, function (result) {console.log("resolve promise with:", result)}); return {then : function (fn) {}}\n'
                    +  '    }\n'
    }

    //  TODO: client side logging and error handling...

    var serialized = 'serialized = ' + clientized.toString()
        .replace('ENVIRONMENT', environment)
        .replace('CLIENT', client_fn.toString());

    eval(serialized);

    return serialized;
}
